articles/contributing/api-design-principles.md

```diff
@@ -0,0 +1,426 @@
+        subroutine to be called in more places (e.g.: inside of
+        other functions), and provides useful optimization
+        information to the compiler that can affect performance and
+        optimizations.
+
+    - `ForEach<'TInput, 'TOutput>('TInput => 'TOutput, 'TInput[]) => 'TOutput[]`
+        and `Mapped<'TInput, 'TOutput>('TInput -> 'TOutput, 'TInput[]) -> 'TOutput[]`
+        differ in the guarantees made with respect to
+        determinism; both are useful in different circumstances.
+
+    - API routines that transform the application of quantum
+        operations can often be carried out in a deterministic
+        fashion and hence can be made available as functions such as
+        `CControlled<'T>(op : 'T => Unit) => ((Bool, 'T) => Unit)`.
+
+    - **DO** generalize the input type as much as reasonable for each
+        function and operation, using type parameters as needed. For
+        example:
+      - `ApplyToEach` has type `<'T>(('T => Unit), 'T[]) => Unit`
+          rather than the specific type of its most common
+          application, `((Qubit => Unit), Qubit[]) => Unit`.
+
+- **DO** choose input and output types for functions and operations
+    that are predictable, and that communicate the purpose of a
+    callable.
+
+  - **DO** use tuple types to logically group inputs and outputs
+      that are only significant when considered together. Consider
+      using a user-defined type in these cases. For example:
+
+    - A function to output the local minima of another function
+        may need to take bounds of a search interval as input, such
+        that `LocalMinima(fn : (Double -> Double), (left : Double, right : Double)) : Double`
+        may be an appropriate signature.
+
+    - An operation to estimate a derivative of a machine learning
+        classifier using the parameter shift technique may need to
+        take both the shifted and unshifted parameter vectors as
+        inputs. An input similar to
+        `(unshifted : Double[], shifted : Double[])` may be appropriate in
+        this case.
+
+  - **DO** order items in input and output tuples consistently
+      across different functions and operations. For example:
+
+    - If considering two or functions or operations that each take
+        a rotation angle and a target qubit as inputs, ensure that
+        they are ordered the same in each input tuple. That is,
+        prefer `ApplyRotation(angle : Double, target : Qubit) : Unit is Adj + Ctl`
+        and `DelayedRotation(angle : Double, target : Qubit) : (Unit => Unit is Adj + Ctl)`
+        to `ApplyRotation(target : Qubit, angle : Double) : Unit is Adj + Ctl`
+        and `DelayedRotation(angle : Double, target : Qubit) : (Unit => Unit is Adj + Ctl)`.
+
+- **DO** design functions and operations to work well with Q\#
+    language features such as partial application.
+
+  - **DO** order items in input tuples such that the most commonly
```

Comment:
this seems duplicated (or a concrete example) of *design functions and operations to compose well with other*

Reply:
Used new examples to clarify that the former is about composability with other functions and operations while this one is about composability with language features.